src/Confluent.Kafka/IProducer.cs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2018 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Confluent.Kafka.Impl;
+using Confluent.Kafka.Internal;
+using Confluent.Kafka.Serialization;
+using System.Collections.Concurrent;
+
+
+namespace Confluent.Kafka
+{
+    /// <summary>
+    ///     Defines a high-level Apache Kafka producer client (without 
+    ///     serialization capability).
+    /// </summary>
+    public interface IProducer : IDisposable
+    {
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Poll_int"]/*' />
+        int Poll(int millisecondsTimeout);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Poll_TimeSpan"]/*' />
+        int Poll(TimeSpan timeout);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Poll"]/*' />
+        [Obsolete("Use an overload of Poll with a finite timeout.", false)]
+        int Poll();
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="GetSerializingProducer"]/*' />
+        ISerializingProducer<TKey, TValue> GetSerializingProducer<TKey, TValue>(ISerializer<TKey> keySerializer, ISerializer<TValue> valueSerializer);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Message"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(Message message);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_TKey_TValue"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(string topic, byte[] key, byte[] val);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_Partition_TKey_TValue_Timestamp_IEnumerable"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(
+            string topic, Partition partition, 
+            byte[] key, byte[] val, 
+            Timestamp timestamp, 
```

Comment:
Although not all values of `Timestamp` are valid here (i.e. only timestamps of type CreationTime are allowed), I think it makes the most sense to use this type here (and not `DateTime` or `DateTimeOffset`) since. 1. consistency with `Message` parameter variant of this method. 2. `Timestamp` encapsulates the -> unix timestamp functionality already. 3. General screwed up-ness of `DateTime` and non-universal usage of `DateTimeOffset`.

Reply:
There are `Timestamp` constructors that take a `DateTime` or `DateTimeOffset` (which effectively encapsulates the conversion to unix time).